File: Week_08/lru_cache/146_lru_cache.java
class LRUCache {
    private HashMap<Integer, DLinkedNode> cache;
    private DLinkedNode head;
    private DLinkedNode tail;
    private int size;
    private int capacity;

    public LRUCache(int capacity) {
        this.cache = new HashMap<>(capacity);
        this.head = new DLinkedNode();
        this.tail = new DLinkedNode();
        this.size = 0;
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        DLinkedNode node = cache.get(key);
        if(node == null) {
            return -1;
        }
        moveToHead(node);
        return node.value;
    }

    public void put(int key, int value) {
        DLinkedNode node = cache.get(key);

        if(node == null) {
            DLinkedNode newNode = new DLinkedNode(key, value);
            cache.put(key, newNode);
            addToHead(newNode);
            size ++;
            if(size > capacity) {
                DLinkedNode tail = removeTail();
                cache.remove(tail.key);
            }
        } else {
            node.value = value;
            moveToHead(node);
        }
    }

    private void addToHead(DLinkedNode node) {
        DLinkedNode next = head.next;
        head.next = node;
        node.next = next;
        node.prev = head;
        next.prev = node;
    }

    private void moveToHead(DLinkedNode node) {
        removeNode(node);
        addToHead(node);
    }

    private void removeNode(DLinkedNode node) {
        DLinkedNode prev = node.prev;
        DLinkedNode next = node.next;
        prev.next = next;
        next.prev = prev;
    }

    private DLinkedNode removeTail() {
        DLinkedNode res = tail.prev;
        removeNode(res);
        return res;
    }
}

class DLinkedNode{
    DLinkedNode prev;
    DLinkedNode next;
    int key;
    int value;
    public DLinkedNode() {}
    public DLinkedNode(int key, int value) {
        this.key = key;
        this.value = value;
    }

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: Week_02/python/146_lru_cache.py
class DLinkedNode:
    def __init__(self, key=None, value=None):
        self.pre = None
        self.nxt = None
        self.key = key
        self.value = value


class LRUCache:
    def __init__(self, capacity: int):
        self.cache = {}
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.size = 0
        self.capacity = capacity
        self.head.nxt = self.tail
        self.tail.pre = self.head

    def get(self, key: int) -> int:
        node = self.cache.get(key)
        if not node:
            return -1

        self._move_to_head(node)

        return node.value

    def put(self, key: int, value: int) -> None:
        node = self.cache.get(key)
        if not node:
            new_node = DLinkedNode(key, value)
            self.cache[key] = new_node
            self._add_to_head(new_node)
            self.size += 1
            if self.size > self.capacity:
                tail = self._remove_tail()
                self.cache[tail.key] = None
                self.size -= 1

            return

        node.value = value
        self._move_to_head(node)

    def _move_to_head(self, node: DLinkedNode):
        self._remove_node(node)
        self._add_to_head(node)

    def _add_to_head(self, node: DLinkedNode):
        nxt = self.head.nxt
        self.head.nxt = node
        node.pre = self.head
        node.nxt = nxt
        nxt.pre = node

    def _remove_node(self, node: DLinkedNode):
        nxt = node.nxt
        pre = node.pre

        nxt.pre = pre
        pre.nxt = nxt

    def _remove_tail(self) -> DLinkedNode:
        node = self.tail.pre
        self._remove_node(node)
        return node

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
